### PR TITLE
Final adjustments and fixes on New Program Engagement modal

### PR DESCRIPTION
**Pull request by jjbennett**
# Critical Changes

# Changes

# Issues Closed

# New Metadata

# Deleted Metadata

# Definition of Done
  Refer to [Asteroids DoD document](https://salesforce.quip.com/iq2mAy4i62oM) to see any additional details for the items below
~~Any net new LWC work has Sa11y tests & 50% or above lines JEST test coverage~~ 
~~CRUD/FLS is enforced in Apex~~
~~Permission sets are updated to account for CRUD|FLS|Tab|Classes~~
~~Field sets are updated to account for new fields~~
- [x] UX approval or UX not necessary
- [x] Link the pull request and work item by PR comment and Chatter post respectively, e.g. GUS: W-0000000: Work Name (https://github.com/SalesforceFoundation/PMM/pull/303)
- [ ] All **acceptance criteria** have been met
    - [x] Developer
    - [ ] Code Reviewer
~~PR contains draft release notes~~
- [ ] QE story level testing completed

### CODE REVIEW

File: force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js [hunk 1]
```diff
@@ -155,6 +179,15 @@ export default class NewProgramEngagement extends LightningElement {
         }
     }
 
+    handleFormChange() {
+        if (!this.hasError) {
+            return;
+        }
+
+        this.form.querySelector("lightning-messages").setError(undefined);
```
Comment (bmuniswamy) [inline force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js:187]: Do we need to set this to undefined wouldn't record-edit form clear the value when there is no error?

File: force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js [hunk 2]
```diff
@@ -262,6 +304,10 @@ export default class NewProgramEngagement extends LightningElement {
         this.showEngagementForm = true;
     }
 
+    get outerClass() {
+        return this.newContactMode ? "outer contact" : "outer";
```
Comment (bmuniswamy) [inline force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js:308]: It could be an overkill but do these need to be constants

**Review by bmuniswamy — COMMENTED** [review]
Have a couple of comments.Still need to play with the UI and review the code again

File: force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js (hunk 1 above)
Comment (jjbennett) [reply force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js:187]: No the error does not get cleared until submit is hit again. This will clear the error as soon as the user starts interacting with the form again.

File: force-app/main/default/classes/ProgramController.cls [hunk 3]
```diff
@@ -1,11 +1,40 @@
...
+    private static final String ENGAGEMENT_FIELD_SET = 'CreateProgramEngagement';
+    private static final String CONTACT_FIELD_SET = 'CreateContact';
+    private static final String PROGRAM_COHORTS_KEY = 'programCohorts';
+    private static final String ENGAGEMENT_FIELD_SET_KEY = 'engagementFieldSet';
+    private static final String CONTACT_FIELD_SET_KEY = 'contactFieldSet';
 
     @AuraEnabled(cacheable=true)
-    public static List<ProgramCohort__c> getProgramCohortsFromProgramId(Id programId) {
+    public static Map<String, Object> getNewEngagementSetup(Id programId) {
+        Map<String, Object> newEngagementSetup = new Map<String, Object>();
```
[the remark this answers is not in the document]
Comment (jjbennett) [reply force-app/main/default/classes/ProgramController.cls:14]: Ahh yes, I was probably being lazy. I will fix that up.

[the remark this answers is not in the document]
Comment (jjbennett) [reply force-app/main/default/classes/ProgramController.cls:14]: @bmuniswamy - ok so I started working on this but now I'm remembering why I chose to leave it in the controller. From my understanding the controller knows what the UI wants and the Service is decoupled from the UI. I think this is ok because it is the controller just mapping the data returned to the UI... @AndersonTarren @ptewson-sfdo thoughts?

[the remark this answers is not in the document]
Comment (ptewson-sfdo) [reply force-app/main/default/classes/ProgramController.cls:14]: I'm still diving into this and haven't wrapped my head around all the aspects of this change yet, but I am leaning towards Jenny's perspective - if this arrangement of this map is in service of the needs of the UI, it probably belongs in the controller. 

There's maybe an argument here to move this even further into a view-specific structure with defined behaviors and properties set up for the UI/view itself, but I'm not going to advocate for that without a better understanding of all these updates yet! 

TL;DR I think it probably belongs in the controller, but maybe hinges on the nuance of whether this is intrinsic to how we retrieve the data from the database, or an arrangement that serves the UI.

**Comment by jjbennett** [comment]
[W-9819240](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE000001fMVUYA2/view)

File: force-app/main/default/classes/ProgramController.cls (hunk 3 above)
[the remark this answers is not in the document]
Comment (ptewson-sfdo) [reply force-app/main/default/classes/ProgramController.cls:14]: And now I'm reading through more, I think adding more structure to this response might be worth considering - a class dedicated to storing all this information that can be handed back to the UI layer, with properties for cohorts, engagement fields, and contact fields. We might be able to define some other behaviors in there as well - canUpdateContact() or canUpdateProgram() - that might help us keep the design cleaner? I'm still getting my head around the ramifications of all that though. Happy to do a post-lunch (Pac time) call to discuss!

File: force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js [hunk 4]
```diff
@@ -69,44 +81,64 @@ export default class NewProgramEngagement extends LightningElement {
...
+    @wire(getPicklistValues, {
+        recordTypeId: "$recordTypeId",
+        fieldApiName: STAGE_FIELD,
     })
-    wiredCohorts({ error, data }) {
-        if (data) {
-            this.cohorts = data;
-            this.setCohortOptions();
+    wiredStageValues({ error, data }) {
+        if (data && data.values) {
```
Comment (bmuniswamy) [inline force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js:126]: @jjbennett, Do we also need to check if data.defaultValue is there. I am seeing an error when trying to create new PE from BSDT using individual option. I am seeing an empty modal

Screen Shot 2021-09-07 at 3 05 42 PM

File: force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js [hunk 5]
```diff
@@ -31,8 +32,14 @@ import newContact from "@salesforce/label/c.New_Contact";
 import cancelAndBack from "@salesforce/label/c.Cancel_and_Back";
 import cantFindContact from "@salesforce/label/c.Cant_Find_Contact";
 
-const CREATE_PROGRAM_ENGAGEMENT_FIELD_SET = "CreateProgramEngagement";
-const CREATE_CONTACT_FIELD_SET = "CreateContact";
+const ACTIVE = "Active";
+const ENROLLED = "Enrolled";
+const ALLOWED_STAGES = [ACTIVE, ENROLLED];
```
Comment (ptewson-sfdo) [inline force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js:37]: Could this be set up as a enum-like structure? 

```
const ACTIVE_ENGAGEMENT_STAGES_ENUM = Object.freeze({
    ACTIVE:  "Active",
    ENROLLED: "Enrolled"
});
```

File: force-app/main/default/classes/ProgramController.cls [hunk 6]
```diff
@@ -1,6 +1,13 @@
 public with sharing class ProgramController {
     @TestVisible
     private static ProgramService service = new ProgramService();
+    @TestVisible
+    private static FieldSetService fieldSetService = new FieldSetService();
+    private static final String ENGAGEMENT_FIELD_SET = 'CreateProgramEngagement';
+    private static final String CONTACT_FIELD_SET = 'CreateContact';
+    private static final String PROGRAM_COHORTS_KEY = 'programCohorts';
```
Comment (ptewson-sfdo) [inline force-app/main/default/classes/ProgramController.cls:8]: I think maybe the PROGRAM_COHORTS_KEY is no longer needed with the (totally rad) updates?

File: force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js (hunk 5 above)
Comment (jjbennett) [reply force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js:38]: yea but thinking it's not really necessary here since this is temporary?

File: force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js (hunk 4 above)
Comment (jjbennett) [reply force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js:127]: yes yes we do, thanks!

File: force-app/main/default/classes/ProgramController.cls (hunk 3 above)
[the remark this answers is not in the document]
Comment (jjbennett) [reply force-app/main/default/classes/ProgramController.cls]: I believe we are relying on Salesforce to handle perms in this situation by using the record-edit-form.

File: force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js (hunk 5 above)
Comment (ptewson-sfdo) [reply force-app/main/default/lwc/newProgramEngagement/newProgramEngagement.js:38]: That makes perfect sense, good perspective!

**Review by ptewson-sfdo — APPROVED** [review]
Looks good, and nice persistence getting this polished up!
